README.md

```diff
@@ -450,6 +454,65 @@ stdio to an in-memory [virtual terminal](https://github.com/hinshun/vt10x).
 
 For some examples, you can see any of the tests in this repo.
 
+### Using a mock
+
+Instead of calling the survey functions directly, you can create a survey struct and call the functions from there.
+
+```golang
+
+survey := survey.Survey{}
+
+response := false
+prompt := &survey.Confirm{
+    Message: "Do you like pie?",
+}
+survey.AskOne(prompt, &response)
+
+```
+
+If you create only one survey struct at the top level of your program and pass it to all functions, you can test those functions by replacing the struct with a mock provided by survey.
+
+#### main
+
+```golang
+func main() {
+    survey := survey.Survey{}
+
+    AskForPie(survey)
+}
+
+func AskForPie(survey survey.SurveyInterface) bool {
+    response := false
+    prompt := &survey.Confirm{
+        Message: "Do you like pie?",
+    }
+    survey.AskOne(prompt, &response)
+
+    return response
+}
+
+```
+
+#### Test
+
+```golang
+func TestAskForPie(t *testing.T) {
+    
+    //create mock
+    mock := survey.SurveyMock{} 
+    //set the response the "user" should select 
+    mock.setResponse(true)
```

Comment:
If AskForPie is going to call `survey.Ask()` multiple times, how would we mock all those using `SetResponse`?

Reply:
The current implementation does not allow to set multiple responses and would always use the last one that is set. 
However, I feel like this could be acceptable behaviour, because it forces the developers to separate different `ask` calls into multiple functions that only ever do one thing (as they should 😉)

Alternatively, `setResponse()` could return the struct and store the responses in a slice. This would make this call joinable and allow multiple answers.

Reply:
I feel strongly that the mock solution should be flexible enough to mock all responses, no matter how many times Ask or AskOne has been called. I may be able to contribute this improvement to your mock implementation.

Example use case:
```
surveyer := survey.SurveyMock{}
surveyer.SetReponse(map[string]interface{}{
  "name": "foo",
  "message": "hello",
})
// ...
DoSomething(surveyer)
```
This test setup will work as long as DoSomething is calling Ask with two questions. However, let's say that the implementation of DoSomething needs to change because the prompt of question 2 needs to dynamically adjust to the response from question 1. The way to do that with survey is to switch from Ask to 2x AskOne. After the implementation changes in this way, the current test from the example will break. However, if the mock solution does not support stubbing several AskOnes, then the user will be forced to break up DoSomething into separate functions just to appease the mock implementation. I think this would cause frustration to the user.

Reply:
This is a usecase i did not think of before.

How would you feel about something like this (inspired by [jest](https://jestjs.io/docs/mock-functions#mock-return-values)):
```golang

mock := survey.Surveyor{}
mock.SetResponseOnce("first").SetResponseOnce("second").SetRespose("all others")

DoSomething(surveyor)

```
Now the mock would respond to the first question inside `DoSomething` with "first", two the second with "second" and to all other questions with "all others". 

Now if `DoSomething` is using `Ask`, it would just fill the question struct with the answers "first", "second", "all others", "all others" if the prompt had four questions. 

If `DoSomething` is using `AskOne`, it would respond to the first `AskOne` with "first", and to the second with "second", etc...

Reply:
I like your proposal; it looks more flexible already! One thing that I still think would be great if SurveyMock could do is verify that the prompt was shown to the user with the right text, since this text is the most important part of how the user interacts with Survey. [We do it like this](https://github.com/cli/cli/blob/7d0b37d8f37b554e3acbc4d00c72d24e08be9e90/pkg/cmd/pr/merge/merge_test.go#L1103-L1106), which feels similar to what you've proposed, but allows for a greater level of precision.

Reply:
I like that way. Can you implement it?

Reply:
Yeah! I'm thinking to build open your PR later this week.

Reply:
Would you merge the PR if I'd clean up the documentation and mark the (new) testing part with a disclaimer that the mock API is still subject to change?

I would like to use the current state in my project and I don't mind having to redo some tests with the new and improved API.

Reply:
I think if we're going down this path we probably also want to provide methods on the mock for setting a specific field with a value (something like `.setFieldResponse("name", "SirRegion")`. I'm not sure there isn't a lot of value in coupling a test to the order in which the questions are asked but I think if someone wants to structure there test that way, we should support them

Reply:
@SirRegion Yes, let's undo the overall README changes for now (to keep things as they were before) and have there just be a small mention & example of the mock functionality.

@AlecAivazis How would the `SetFieldResponse()` functionality allow for asserting the order in which questions were asked?

Reply:
I mean to have both methods so a user could decide if they want to test the order of responses or just if a response was given for a field.

Reply:
@AlecAivazis Agreed! It's probably better to be strict with verification out-of-the-box (i.e. to verify the order of prompts by default) but give the developer a chance to opt out if needed.

Reply:
@mislav I reverted the commit that changed the readme for now and updated the Testing part to include a warning about the unstable mock API.